staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -15,58 +15,84 @@
-			continue
+	if s.XIntOrString {
+		return NewObjectType("intOrString", map[string]*DeclField{
+			"strVal": {Name: "strVal", Type: StringType},
+			"intVal": {Name: "intVal", Type: IntType},
+		})
+	}
+	if s.XEmbeddedResource {
+		return NewObjectType("embedded", map[string]*DeclField{
+			"kind":       {Name: "kind", Type: StringType},
+			"apiVersion": {Name: "apiVersion", Type: StringType},
+			"metadata": {Name: "metadata", Type: NewObjectType("metadata", map[string]*DeclField{
+				"name":         {Name: "name", Type: StringType},
+				"generateName": {Name: "generateName", Type: StringType, Required: false},
+			})},
+			"spec":   {Name: "spec", Type: DynType},
+			"status": {Name: "status", Type: DynType, Required: false},
+		})
+	}
+	if s.XPreserveUnknownFields {
+		return DynType
+	}
+	switch declType.TypeName() {
+	case ListType.TypeName():
+		return NewListType(SchemaDeclType(s.Items))
+	case MapType.TypeName():
+		if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+			return NewMapType(StringType, SchemaDeclType(s.AdditionalProperties.Structural))
 		}
-		switch declType.TypeName() {
-		case ListType.TypeName():
-			if s.Items.Len() == 1 {
-				return NewListType(SchemaDeclType(s.Items.Schema))
-			}
-		case MapType.TypeName():
-			if s.AdditionalProperties != nil {
-				return NewMapType(StringType, SchemaDeclType(s.AdditionalProperties.Schema))
-			}
-			fields := make(map[string]*DeclField, len(s.Properties))
-			required := make(map[string]struct{}, len(s.Required))
-			for _, name := range s.Required {
-				required[name] = struct{}{}
+		fields := make(map[string]*DeclField, len(s.Properties))
+
+		required := map[string]bool{}
+		if s.ValueValidation != nil {
+			for _, f := range s.ValueValidation.Required {
+				required[f] = true
 			}
-			for name, prop := range s.Properties {
-				_, isReq := required[name]
-				fields[name] = &DeclField{
-					Name:         name,
-					Required:     isReq,
-					Type:         SchemaDeclType(&prop),
-					defaultValue: prop.Default,
-					enumValues:   prop.Enum,
+		}
+		for name, prop := range s.Properties {
+			var enumValues []interface{}
+			if prop.ValueValidation != nil {
+				for _, e := range prop.ValueValidation.Enum {
+					enumValues = append(enumValues, e.Object)
 				}
 			}
-			return NewObjectType("object", fields)
-		case StringType.TypeName():
-			switch s.Format {
-			case "byte", "binary":
+			fields[Escape(name)] = &DeclField{
+				Name:         Escape(name),
+				Required:     required[name],
+				Type:         SchemaDeclType(&prop),
+				defaultValue: prop.Default.Object,
+				enumValues:   enumValues, // TODO: escape?
+			}
+		}
+		return NewObjectType("object", fields)
+	case StringType.TypeName():
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "byte":
+				return StringType // OpenAPIv3 byte format represents base64 encoded string
+			case "binary":
 				return BytesType
 			case "google-duration":
```

Comment:
is google-duration a thing in kube?

Reply:
Looks like the official list of formats we use is from kube-openapi:
- https://github.com/kubernetes/kube-openapi/blob/1a6458611d189dc17e98a0824dc92536365efedf/pkg/validation/validate/type.go
- https://github.com/kubernetes/kube-openapi/blob/1a6458611d189dc17e98a0824dc92536365efedf/pkg/validation/strfmt/format.go#L105

I'll remove `google-*` and add some test for the date/time formats to make sure they're working as expected.